paimon-core/src/main/java/org/apache/paimon/table/source/snapshot/StartingScanner.java

```diff
@@ -22,38 +22,74 @@
 import org.apache.paimon.table.source.TableScan;
 import org.apache.paimon.utils.SnapshotManager;
 
-import javax.annotation.Nullable;
-
 import java.util.Collections;
 import java.util.List;
 
 /** Helper class for the first planning of {@link TableScan}. */
 public interface StartingScanner {
 
-    @Nullable
     Result scan(SnapshotManager snapshotManager, SnapshotSplitReader snapshotSplitReader);
 
     /** Scan result of {@link #scan}. */
-    class Result {
+    interface Result {
+        long snapshotId();
+
+        List<DataSplit> splits();
+    }
+
+    /** A null Result. */
+    class NullResult implements Result {
+        @Override
+        public long snapshotId() {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public List<DataSplit> splits() {
+            throw new UnsupportedOperationException();
+        }
+    }
 
+    /** A non null Result. */
+    abstract class NonNullResult implements Result {
         private final long snapshotId;
         private final List<DataSplit> splits;
 
-        public Result(long snapshotId) {
+        public NonNullResult(long snapshotId) {
             this(snapshotId, Collections.emptyList());
         }
 
-        public Result(long snapshotId, List<DataSplit> splits) {
+        public NonNullResult(long snapshotId, List<DataSplit> splits) {
             this.snapshotId = snapshotId;
             this.splits = splits;
         }
 
+        @Override
         public long snapshotId() {
             return snapshotId;
         }
 
+        @Override
         public List<DataSplit> splits() {
             return splits;
         }
     }
+
+    /** Contains an existing snapshot. */
+    class ExistingSnapshotResult extends NonNullResult {
```

Comment:
`ScannedResult`: Result with scanned snapshot. Next snapshot should be current snapshot plus 1.